aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp

```diff
@@ -2568,6 +2568,225 @@ void dequantize_tensor_per_tensor_affine_cpu(
+#if defined(__ARM_NEON__) || defined(__aarch64__)
+// Generic template defaults to naive quantize implementation
+template <typename T>
+void quantize_tensor_per_channel_arm(
+    const float* in,
+    Tensor qtensor,
+    const double* scales,
+    const int64_t* zero_points,
+    const bool channels_last,
+    const int64_t batches,
+    const int64_t elements_per_channel,
+    const int64_t channels) {
+  auto out = qtensor.data_ptr<T>();
+  if (channels_last) {
+    for (auto b = 0; b < batches; ++b) {
+      for (auto e = 0; e < elements_per_channel; ++e) {
+        for (auto c = 0; c < channels; ++c) {
+          auto i = b * channels * elements_per_channel + e * channels + c;
+          out[i] =
+              at::native::quantize_val<T>(scales[c], zero_points[c], in[i]);
+        }
+      }
+    }
+  } else {
+    for (auto b = 0; b < batches; ++b) {
+      for (auto c = 0; c < channels; ++c) {
+        for (auto e = 0; e < elements_per_channel; ++e) {
+          auto i = b * channels * elements_per_channel +
+              c * elements_per_channel + e;
+          out[i] =
+              at::native::quantize_val<T>(scales[c], zero_points[c], in[i]);
+        }
+      }
+    }
+  }
+}
+
+// Specialized implementation from caffe2::Int8Quantize.
+// There may be slight accuracy difference between this and implementation of
+// quantize_val
+// TODO Update quantize_tensor_per_channel_arm implementation to follow
+// quantize_val, i.e. f = Round(value/scale + zero_point)
+// TODO Make quantize_tensor_per_channel_arm work for other datatypes too (int8,
+// int32).
+template <>
+void quantize_tensor_per_channel_arm<c10::quint8>(
+    const float* in,
+    Tensor qtensor,
+    const double* scales,
+    const int64_t* zero_points,
+    const bool channels_last,
+    const int64_t batches,
+    const int64_t elements_per_channel,
+    const int64_t channels) {
+  auto out = (uint8_t*)qtensor.data_ptr<c10::quint8>();
+#if defined(__ARM_NEON__)
+  // magic float and magic int to take care of rounding
+  // int magic_round(float f): interpret_int32(f + 12582912.0f) - 0x4B400000
+  // Some detail:
+  // 12582912.0f is 2**23 + 2**22. The trick is based on the fact that when you
+  // add a small number to a large number, the result rounds to the precision of
+  // the least significant bit of the large number. For IEEE-754
+  // single-precision number mantissa has 23 bits, and adding 2**23 would cause
+  // rounding to the nearest even integer. The we cast to int and subtract the
+  // same number (0x4B400000 is the integer representation of 12582912.0f) to
+  // get only the mantissa. This works if -2**22 < x < 2**22, but preserves the
+  // sign for negative numbers.
+  const float32x4_t vmagic_float = vdupq_n_f32(12582912.0f);
+  const int32x4_t vmagic_int = vdupq_n_s32(0x4B400000);
+  // Copy zero_points (int64_t) into int32_t array
+  // Copy reciprocal of scales (double) into float array
+  int32_t zero_points_int32t[channels];
```

Comment:
Why not use std::vector?

Reply:
I am honestly not familiar this way of allocating dynamic array.

Reply:
This is a static array, since we know that zero points should have exactly `channels` number of elements.

Reply:
Yes but channels is not known at compile time, right?

Reply:
Hmm looked into it and it looks like the runtime difference won't really be any different. `std::vector` might use a little more memory, but probably negligible enough to just pick whichever is easier to read.

I'll switch over to using `std::vector`, since the current route above requires you to free the memory pointer manually.

Reply:
That would be a very strong reason to do it.